url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/RuleCollection.java

```diff
@@ -0,0 +1,862 @@
+							{
+								AbstractRule a = i.next();
+								if(a.getCondition().get(d.getID()).get(v.getID()).isFullyActivated())
+								{
+									i.remove();
+								}
+							}
+						}
+					}
+				}
+				coveredByFlag = true;
+			}
+			
+			if(candidates.size() > 0)
+			{
+				for (Iterator <AbstractRule> i = candidates.iterator(); i.hasNext();)
+				{
+					AbstractRule parent = i.next();
+					if(!parent.getCondition().equalsKeys(R.getCondition()))
+						i.remove();
+				}
+				
+				for (AbstractRule a : candidates)
+				{
+					if(a.equals(R))
+					{
+						return false;
+					}
+				}
+				
+				if(candidates.size() > 0)
+				{
+					boolean success = Children.add(R);
+					AbstractRule r = R;
+					if(!success)
+						r = Children.get(R.getCondition(), R.getAction());
+					
+					for (AbstractRule parent : candidates)
+					{
+						if(parent.getChildren() == null)
+							parent.initChildren();
+						parent.getChildren().add(r);
+					}
+				}
+				else
+				{
+					attachRule(R);
+				}
+			}
+			else 	
+			{
+				attachRule(R);
+				
+				for(AbstractRule a : coveredBy)
+				{
+					if(!a.equals(R))
+					{
+						remove(a);
+						Children.add(a);
+					}
+				}
+				
+				HashMap<Object, HashMap<Object,LinkedList <AbstractRule>>> crc = Children.get(R.getAction().getID());
+				if(crc != null)
+				{
+					HashSet <AbstractRule> covers = new HashSet<AbstractRule> ();
+					for(Dimension d : R.getCondition().values())
+					{
+						HashMap<Object,LinkedList <AbstractRule>> crrc = crc.get(d.getID());
+						if(crrc != null)
+						{
+							for(Value v : d.values())
+							{
+								if(v.isFullyActivated())
+								{
+									LinkedList <AbstractRule> crrrc = crrc.get(v.getID());
+									if(crrrc != null)
+										covers.addAll(crrrc);
+								}
+							}
+						}
+					}
+					
+					for(AbstractRule c : covers)
+					{
+						if(R.covers(c) && !R.equals(c))
+						{
+							if(R.getChildren() == null)
+								R.initChildren();
+							R.getChildren().add(c);
+						}
+					}
+				}
+			}
+			R.resetMatchStatistics();
+			return true;
+		}
+	}
+	
+	/**
+	 * Adds all rules in the specified collection of rules into the rule collection. If a rule in the specified 
+	 * collection is equal to any rules currently in this collection, the rule will NOT 
+	 * be added.
+	 * @param R The rules to add to the rule collection.
+	 */
+	public void addAll (Collection <? extends AbstractRule> R)
+	{
+		for(AbstractRule r : R)
+		{
+			add(r);
+		}
+	}
+	
+	/**
+	 * Removes the specified rule from this rule collection if the rule is contained within the rule collection. This 
+	 * method also removes all of the rule variations associated with the rule being removed as long as the variation 
+	 * is not also a variation for a different rule in the rule store.
+	 * @param R The rule you want to remove.
+	 * @return True if the rule was successfully removed, otherwise false.
+	 */
+	public boolean remove (AbstractRule R)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.